src/build/urlfetcher.go

```diff
@@ -0,0 +1,277 @@
+	return id
+}
+
+func isURL(u string) bool {
+	return (7 <= len(u) && u[:7] == "http://") ||
+		(8 <= len(u) && u[:8] == "https://")
+}
+
+func (info *URLInfo) getBlobFileName() (fileName string) {
+	return filepath.Join(info.Fetcher.cacheDir, info.ID[:2], info.ID)
+}
+
+func (info *URLInfo) getInfoFileName() (fileName string) {
+	return info.getBlobFileName() + ".json"
+}
+
+func (info *URLInfo) isEtagValid() bool {
+	if !info.HasEtag {
+		return false
+	}
+
+	var httpClient *http.Client
+	if info.Fetcher.client == nil {
+		httpClient = http.DefaultClient
+	} else {
+		httpClient = info.Fetcher.client
+	}
+
+	response, err := httpClient.Head(info.URL)
+	if err != nil {
+		return false
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || 300 <= response.StatusCode {
+		return false
+	}
+
+	etag, ok := response.Header["Etag"]
+	if ok && 1 <= len(etag) && etag[0] == info.Etag {
+		return true
+	}
+
+	return false
+}
+
+func (info *URLInfo) download() (err error) {
+	log.Infof("Downloading `%s` into `%s`", info.URL, info.FileName)
+
+	var httpClient *http.Client
+	if info.Fetcher.client == nil {
+		httpClient = http.DefaultClient
+	} else {
+		httpClient = info.Fetcher.client
+	}
+
+	response, err := httpClient.Get(info.URL)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || 300 <= response.StatusCode {
+		return fmt.Errorf("Got non-2xx status for `%s`: %s", info.URL, response.Status)
+	}
+
+	err = os.MkdirAll(filepath.Dir(info.FileName), 0755)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Create(info.FileName)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	n, err := io.Copy(f, response.Body)
+	if err != nil {
+		return err
+	}
+
+	info.Size = n
+
+	if etag, ok := response.Header["Etag"]; ok {
+		info.HasEtag = true
+		info.Etag = etag[0]
+	} else {
+		info.HasEtag = false
+		info.Etag = ""
+	}
+
+	err = info.store()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (info *URLInfo) load() (ok bool, err error) {
+	fileName := info.getInfoFileName()
+
+	data, err := ioutil.ReadFile(fileName)
+
+	if err != nil {
+		//if pe, ok0 := err.(*os.PathError); ok0 && pe.Err == os.ErrNotExist {
+		if _, ok0 := err.(*os.PathError); ok0 {
+			return false, nil
+		}
+		return false, fmt.Errorf("Failed to read urlinfo file %s content, error: %s", fileName, err)
+	}
+
+	if err = json.Unmarshal(data, info); err != nil {
+		return false, fmt.Errorf("Failed to parse urlinfo file %s json, error: %s", fileName, err)
+	}
+
+	return true, nil
+}
+
+func (info *URLInfo) store() (err error) {
+	fileName := info.getInfoFileName()
+
+	if err := os.MkdirAll(filepath.Dir(fileName), 0755); err != nil {
+		return err
+	}
+	data, err := json.Marshal(info)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(fileName, data, 0644)
+}
+
+func (info *URLInfo) dump() (data string, err error) {
+	data0, err := json.Marshal(info)
+	return string(data0), err
```

Comment:
You can also use `pretty.Sprint` and/or implement `String` interface for such stuff.